File: Python/0652-m-find-duplicate-subtrees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

"""
Recursion - Divide and Conquer
Time: O(n^2)
Space: O(n^2)
"""
class Solution:
    def findDuplicateSubtrees(self, root: TreeNode) -> List[TreeNode]:
        self.count = collections.defaultdict(int)
        self.res = []
        self.collect(root)
        return self.res

    def collect(self, node):
        if not node:
            return "#"
        serial = "{},{},{}".format(node.val, self.collect(node.left), self.collect(node.right))
        self.count[serial] += 1
        if self.count[serial] == 2:
            self.res.append(node)
        return serial

File: leetcode/Java/0652-m-find-duplicate-subtrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private Map<String, Integer> count;
    private List<TreeNode> res;
    
    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        count = new HashMap<>();
        res = new ArrayList<>();
        collect(root);
        return res;
    }
    
    private String collect(TreeNode node) {
        if (node == null)
            return "#";
        String serial = node.val + "," + collect(node.left) + "," + collect(node.right);
        count.put(serial, count.getOrDefault(serial, 0) + 1);
        if (count.get(serial) == 2)
            res.add(node);
        return serial;
    }
}